connectors/src/connectors/zendesk/lib/zendesk_api.ts

```diff
@@ -1,27 +1,146 @@
+/**
+ * Returns a Zendesk client with the subdomain set to the one in the brand.
+ * Retrieves the brand from the database if it exists, fetches it from the Zendesk API otherwise.
+ */
+export async function changeZendeskClientSubdomain(
+  client: Client,
+  {
+    connectorId = null,
+    brandId,
+  }: {
+    connectorId?: ModelId | null;
+    brandId: number;
+  }
+) {
+  client.config.subdomain = await getZendeskBrandSubdomain(client, {
+    connectorId,
+    brandId,
+  });
+  return client;
+}
+
+/**
+ * Retrieves a brand's subdomain from the database if it exists, fetches it from the Zendesk API otherwise.
+ */
+export async function getZendeskBrandSubdomain(
+  client: Client,
+  {
+    connectorId = null,
+    brandId,
+  }: {
+    connectorId?: ModelId | null;
+    brandId: number;
+  }
+): Promise<string> {
+  if (connectorId) {
+    const brandInDb = await ZendeskBrandResource.fetchByBrandId({
+      connectorId,
+      brandId,
+    });
+    if (brandInDb) {
+      return brandInDb.subdomain;
+    }
+  }
   const {
     result: { brand },
   } = await client.brand.show(brandId);
-  client.config.subdomain = brand.subdomain;
-  return client;
+  return brand.subdomain;
+}
+
+/**
+ * Handles rate limit responses from Zendesk API.
+ * Expects to find the header `Retry-After` in the response.
+ * https://developer.zendesk.com/api-reference/introduction/rate-limits/
+ * @returns true if the rate limit was handled and the request should be retried, false otherwise.
+ */
+async function handleZendeskRateLimit(response: Response): Promise<boolean> {
+  if (response.status === 429) {
+    const retryAfter = Number(response.headers.get("Retry-After")) || 60;
+    await new Promise((resolve) => setTimeout(resolve, retryAfter * 1000));
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Fetches a batch of articles in a category from the Zendesk API.
+ */
+export async function fetchZendeskArticlesInCategory({
+  subdomain,
+  accessToken,
+  categoryId,
+  pageSize = 100,
+  cursor = null,
+}: {
+  subdomain: string;
+  accessToken: string;
+  categoryId: number;
+  pageSize?: number;
+  cursor?: string | null;
+}): Promise<{
+  articles: ZendeskFetchedArticle[];
+  meta: { hasMore: boolean; after_cursor: string };
+}> {
+  assert(
+    pageSize <= 100,
+    `pageSize must be at most 100 (current value: ${pageSize})` // https://developer.zendesk.com/api-reference/introduction/pagination
+  );
+  const runFetch = async () =>
+    fetch(
+      `https://${subdomain}.zendesk.com/api/v2/help_center/categories/${categoryId}/articles?page[size]=${pageSize}` +
+        (cursor ? `&page[after]=${cursor}` : ""),
+      {
+        method: "GET",
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+          "Content-Type": "application/json",
+        },
+      }
+    );
+
+  let rawResponse = await runFetch();
+  while (await handleZendeskRateLimit(rawResponse)) {
```

Comment:
Do we need a loop? 
If we waited the valid amount of time one time and it's still not good then maybe it's okay to fail so we can investigate?

Reply:
but a failed activity would be much easier to investigate than a stuck one so I am suddenly leaning towards having 1 check only

Reply:
after giving it some thought, I'll give this a max number of retries (something in the lines of 5 retries)